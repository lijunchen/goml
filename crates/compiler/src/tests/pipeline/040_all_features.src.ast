struct Tracker {
    label: string,
    count: Ref[int],
    toggled: Ref[bool],
}

enum Record[T] {
    Value(T),
    Pair(T, T),
    Empty
}

enum Maybe[T] {
    Some(T),
    None
}

trait Describe {
  fn describe(Self) -> string;
}

impl Describe for Tracker {
  fn describe(self: Tracker) -> string {
      let Tracker { label: label, count: count, toggled: toggled } = self in
      let current = ref_get(count) in
      let flag = ref_get(toggled) in
      let with_label = string_add("Tracker(", label) in
      let with_count_label = string_add(with_label, ", count: ") in
      let with_count = string_add(with_count_label, int_to_string(current)) in
      let with_flag_label = string_add(with_count, ", toggled: ") in
      string_add(with_flag_label, string_add(bool_to_string(flag), ")"))
  }
}

impl Describe for Record[int] {
  fn describe(self: Record[int]) -> string {
      match self {
          Value(value) => string_add("Value(", string_add(int_to_string(value), ")")),
          Pair(before, after) => let prefix = string_add("Pair(", int_to_string(before)) in
            string_add(prefix, string_add(", ", string_add(int_to_string(after), ")"))),
          Empty => "Empty",
      }
  }
}

impl Describe for Record[string] {
  fn describe(self: Record[string]) -> string {
      match self {
          Value(text) => string_add("Value(", string_add(text, ")")),
          Pair(before, after) => let prefix = string_add("Pair(", before) in
            string_add(prefix, string_add(", ", string_add(after, ")"))),
          Empty => "Empty",
      }
  }
}

fn choose(flag: bool, when_true: T, when_false: T) -> T {
    if flag {
        when_true
    } else {
        when_false
    }
}

fn map_maybe(value: Maybe[T], f: (T) -> U) -> Maybe[U] {
    match value {
        Some(inner) => Some(f(inner)),
        None => None,
    }
}

fn format_total(total: int) -> string {
    string_add("total: ", int_to_string(total))
}

fn increment(value: int) -> int {
    value + 1
}

fn triple(value: int) -> int {
    value * 3
}

fn pair_join(parts: (string, string)) -> string {
    let (left, right) = parts in
    string_add(left, string_add(" -> ", right))
}

fn run_transforms(value: int, transforms: [(int) -> int; 2]) -> [int; 2] {
    let first = array_get(transforms, 0) in
    let second = array_get(transforms, 1) in
    let first_result = first(value) in
    let second_result = second(first_result) in
    array_set([first_result, value], 1, second_result)
}

fn gather(record: Record[int]) -> Maybe[int] {
    match record {
        Value(value) => Some(value),
        Pair(_, after) => Some(after),
        Empty => None,
    }
}

fn build_counter(label: string, start: int) -> (Tracker, () -> Record[int], (int) -> Record[int], () -> Record[string]) {
    let count = ref(start) in
    let toggled = ref(false) in
    let tracker = Tracker {
        label: label,
        count: count,
        toggled: toggled
    } in
    let snapshot = || Value(ref_get(count)) in
    let bump = |delta: int| let before = ref_get(count) in
    let _ = ref_set(count, before + delta) in
    Pair(before, ref_get(count)) in
    let flip = || let before = ref_get(toggled) in
    let _ = ref_set(toggled, !before) in
    let after = ref_get(toggled) in
    Pair(bool_to_string(before), bool_to_string(after)) in
    (tracker, snapshot, bump, flip)
}

fn main() {
    let (tracker, snapshot, bump, flip) = build_counter("goml", 2) in
    let tracker_info = describe(tracker) in
    let first_record = snapshot() in
    let bumped_record = bump(5) in
    let flipped_record = flip() in
    let maybe_first = gather(first_record) in
    let maybe_second = gather(bumped_record) in
    let chosen = choose(true, maybe_second, maybe_first) in
    let stringified = map_maybe(chosen, format_total) in
    let transforms = [increment, triple] in
    let results = run_transforms(4, transforms) in
    let first_result = array_get(results, 0) in
    let second_result = array_get(results, 1) in
    let order_check = int_less(first_result, second_result) && true in
    let first_text = describe(first_record) in
    let bumped_text = describe(bumped_record) in
    let flipped_text = describe(flipped_record) in
    let summary = match stringified {
        Some(text) => string_add("Snapshot: ", text),
        None => "Snapshot: none",
    } in
    let pair_text = pair_join((int_to_string(first_result), int_to_string(second_result))) in
    let bool_text = bool_to_string(order_check) in
    let _ = string_println(tracker_info) in
    let _ = string_println(first_text) in
    let _ = string_println(bumped_text) in
    let _ = string_println(flipped_text) in
    let _ = string_println(summary) in
    let _ = string_println(pair_text) in
    let _ = string_println(bool_text) in
    ()
}

