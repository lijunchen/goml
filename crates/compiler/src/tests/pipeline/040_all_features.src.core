fn impl_Describe_Tracker_describe(self/0: Tracker) -> string {
  let mtmp0 = self/0 in
  let x1 = Tracker.label(mtmp0) in
  let x2 = Tracker.count(mtmp0) in
  let x3 = Tracker.toggled(mtmp0) in
  let toggled/3 = x3 in
  let count/2 = x2 in
  let label/1 = x1 in
  let current/4 = ref_get(count/2) in
  let flag/5 = ref_get(toggled/3) in
  let with_label/6 = string_add("Tracker(", label/1) in
  let with_count_label/7 = string_add(with_label/6, ", count: ") in
  let with_count/8 = string_add(with_count_label/7, int_to_string(current/4)) in
  let with_flag_label/9 = string_add(with_count/8, ", toggled: ") in
  string_add(with_flag_label/9, string_add(bool_to_string(flag/5), ")"))
}

fn impl_Describe_Record_int_describe(self/10: Record[int]) -> string {
  match self/10 {
    Record::Value(x4) => {
      let x4 = Record::Value._0(self/10) in
      let value/11 = x4 in
      string_add("Value(", string_add(int_to_string(value/11), ")"))
    },
    Record::Pair(x5, x6) => {
      let x5 = Record::Pair._0(self/10) in
      let x6 = Record::Pair._1(self/10) in
      let after/13 = x6 in
      let before/12 = x5 in
      let prefix/14 = string_add("Pair(", int_to_string(before/12)) in
      string_add(prefix/14, string_add(", ", string_add(int_to_string(after/13), ")")))
    },
    Record::Empty => {
      "Empty"
    },
  }
}

fn impl_Describe_Record_string_describe(self/15: Record[string]) -> string {
  match self/15 {
    Record::Value(x7) => {
      let x7 = Record::Value._0(self/15) in
      let text/16 = x7 in
      string_add("Value(", string_add(text/16, ")"))
    },
    Record::Pair(x8, x9) => {
      let x8 = Record::Pair._0(self/15) in
      let x9 = Record::Pair._1(self/15) in
      let after/18 = x9 in
      let before/17 = x8 in
      let prefix/19 = string_add("Pair(", before/17) in
      string_add(prefix/19, string_add(", ", string_add(after/18, ")")))
    },
    Record::Empty => {
      "Empty"
    },
  }
}

fn choose(flag/20: bool, when_true/21: T, when_false/22: T) -> T {
  if flag/20 {
      when_true/21
  } else {
      when_false/22
  }
}

fn map_maybe(value/23: Maybe[T], f/24: (T) -> U) -> Maybe[U] {
  match value/23 {
    Maybe::Some(x10) => {
      let x10 = Maybe::Some._0(value/23) in
      let inner/25 = x10 in
      Maybe::Some(f/24(inner/25))
    },
    Maybe::None => {
      Maybe::None
    },
  }
}

fn format_total(total/26: int) -> string {
  string_add("total: ", int_to_string(total/26))
}

fn increment(value/27: int) -> int {
  int_add(value/27, 1)
}

fn triple(value/28: int) -> int {
  int_mul(value/28, 3)
}

fn pair_join(parts/29: (string, string)) -> string {
  let mtmp11 = parts/29 in
  let x12 = mtmp11.0 in
  let x13 = mtmp11.1 in
  let right/31 = x13 in
  let left/30 = x12 in
  string_add(left/30, string_add(" -> ", right/31))
}

fn run_transforms(value/32: int, transforms/33: [(int) -> int; 2]) -> [int; 2] {
  let first/34 = array_get(transforms/33, 0) in
  let second/35 = array_get(transforms/33, 1) in
  let first_result/36 = first/34(value/32) in
  let second_result/37 = second/35(first_result/36) in
  array_set([first_result/36, value/32], 1, second_result/37)
}

fn gather(record/38: Record[int]) -> Maybe[int] {
  match record/38 {
    Record::Value(x14) => {
      let x14 = Record::Value._0(record/38) in
      let value/39 = x14 in
      Maybe::Some(value/39)
    },
    Record::Pair(x15, x16) => {
      let x15 = Record::Pair._0(record/38) in
      let x16 = Record::Pair._1(record/38) in
      let after/40 = x16 in
      Maybe::Some(after/40)
    },
    Record::Empty => {
      Maybe::None
    },
  }
}

fn build_counter(label/41: string, start/42: int) -> (Tracker, () -> Record[int], (int) -> Record[int], () -> Record[string]) {
  let count/43 = ref(start/42) in
  let toggled/44 = ref(false) in
  let tracker/45 = Tracker { label: label/41, count: count/43, toggled: toggled/44 } in
  let snapshot/46 = || => Record::Value(ref_get(count/43)) in
  let bump/49 = |delta/47: int| => let before/48 = ref_get(count/43) in
  let mtmp17 = ref_set(count/43, int_add(before/48, delta/47)) in
  Record::Pair(before/48, ref_get(count/43)) in
  let flip/52 = || => let before/50 = ref_get(toggled/44) in
  let mtmp18 = ref_set(toggled/44, bool_not(before/50)) in
  let after/51 = ref_get(toggled/44) in
  Record::Pair(bool_to_string(before/50), bool_to_string(after/51)) in
  (tracker/45, snapshot/46, bump/49, flip/52)
}

fn main() -> unit {
  let mtmp19 = build_counter("goml", 2) in
  let x20 = mtmp19.0 in
  let x21 = mtmp19.1 in
  let x22 = mtmp19.2 in
  let x23 = mtmp19.3 in
  let flip/56 = x23 in
  let bump/55 = x22 in
  let snapshot/54 = x21 in
  let tracker/53 = x20 in
  let tracker_info/57 = impl_Describe_Tracker_describe(tracker/53) in
  let first_record/58 = snapshot/54() in
  let bumped_record/59 = bump/55(5) in
  let flipped_record/60 = flip/56() in
  let maybe_first/61 = gather(first_record/58) in
  let maybe_second/62 = gather(bumped_record/59) in
  let chosen/63 = choose(true, maybe_second/62, maybe_first/61) in
  let stringified/64 = map_maybe(chosen/63, format_total) in
  let transforms/65 = [increment, triple] in
  let results/66 = run_transforms(4, transforms/65) in
  let first_result/67 = array_get(results/66, 0) in
  let second_result/68 = array_get(results/66, 1) in
  let order_check/69 = bool_and(int_less(first_result/67, second_result/68), true) in
  let first_text/70 = impl_Describe_Record_int_describe(first_record/58) in
  let bumped_text/71 = impl_Describe_Record_int_describe(bumped_record/59) in
  let flipped_text/72 = impl_Describe_Record_string_describe(flipped_record/60) in
  let summary/74 = match stringified/64 {
    Maybe::Some(x24) => {
      let x24 = Maybe::Some._0(stringified/64) in
      let text/73 = x24 in
      string_add("Snapshot: ", text/73)
    },
    Maybe::None => {
      "Snapshot: none"
    },
  } in
  let pair_text/75 = pair_join((int_to_string(first_result/67), int_to_string(second_result/68))) in
  let bool_text/76 = bool_to_string(order_check/69) in
  let mtmp25 = string_println(tracker_info/57) in
  let mtmp26 = string_println(first_text/70) in
  let mtmp27 = string_println(bumped_text/71) in
  let mtmp28 = string_println(flipped_text/72) in
  let mtmp29 = string_println(summary/74) in
  let mtmp30 = string_println(pair_text/75) in
  let mtmp31 = string_println(bool_text/76) in
  ()
}