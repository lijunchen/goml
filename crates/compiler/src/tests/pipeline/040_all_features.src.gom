package main

import (
    "fmt"
)

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func bool_not(x bool) bool {
    return !x
}

func bool_and(x bool, y bool) bool {
    return x && y
}

func int_add(x int, y int) int {
    return x + y
}

func int_mul(x int, y int) int {
    return x * y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_add(a string, b string) string {
    return a + b
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func array_get__Array_2_Fn_int_to_int(arr [2]func(int) int, index int) func(int) int {
    return arr[index]
}

func array_get__Array_2_int(arr [2]int, index int) int {
    return arr[index]
}

func array_set__Array_2_int(arr [2]int, index int, value int) [2]int {
    arr[index] = value
    return arr
}

type ref_int_x struct {
    value int
}

func ref__Ref_int(value int) *ref_int_x {
    return &ref_int_x{
        value: value,
    }
}

func ref_get__Ref_int(reference *ref_int_x) int {
    return reference.value
}

func ref_set__Ref_int(reference *ref_int_x, value int) struct{} {
    reference.value = value
    return struct{}{}
}

type ref_bool_x struct {
    value bool
}

func ref__Ref_bool(value bool) *ref_bool_x {
    return &ref_bool_x{
        value: value,
    }
}

func ref_get__Ref_bool(reference *ref_bool_x) bool {
    return reference.value
}

func ref_set__Ref_bool(reference *ref_bool_x, value bool) struct{} {
    reference.value = value
    return struct{}{}
}

type Tuple2_string_string struct {
    _0 string
    _1 string
}

type Tuple4_Tracker_closure_env_snapshot_0_closure_env_bump_1_closure_env_flip_2 struct {
    _0 Tracker
    _1 closure_env_snapshot_0
    _2 closure_env_bump_1
    _3 closure_env_flip_2
}

type Tracker struct {
    label string
    count *ref_int_x
    toggled *ref_bool_x
}

type closure_env_snapshot_0 struct {
    count_0 *ref_int_x
}

type closure_env_bump_1 struct {
    count_0 *ref_int_x
}

type closure_env_flip_2 struct {
    toggled_0 *ref_bool_x
}

type Record__int interface {
    isRecord__int()
}

type Record__int_Value struct {
    _0 int
}

func (_ Record__int_Value) isRecord__int() {}

type Record__int_Pair struct {
    _0 int
    _1 int
}

func (_ Record__int_Pair) isRecord__int() {}

type Record__int_Empty struct {}

func (_ Record__int_Empty) isRecord__int() {}

type Record__string interface {
    isRecord__string()
}

type Record__string_Value struct {
    _0 string
}

func (_ Record__string_Value) isRecord__string() {}

type Record__string_Pair struct {
    _0 string
    _1 string
}

func (_ Record__string_Pair) isRecord__string() {}

type Record__string_Empty struct {}

func (_ Record__string_Empty) isRecord__string() {}

type Maybe__int interface {
    isMaybe__int()
}

type Maybe__int_Some struct {
    _0 int
}

func (_ Maybe__int_Some) isMaybe__int() {}

type Maybe__int_None struct {}

func (_ Maybe__int_None) isMaybe__int() {}

type Maybe__string interface {
    isMaybe__string()
}

type Maybe__string_Some struct {
    _0 string
}

func (_ Maybe__string_Some) isMaybe__string() {}

type Maybe__string_None struct {}

func (_ Maybe__string_None) isMaybe__string() {}

func impl_Describe_Tracker_describe(self__0 Tracker) string {
    var ret64 string
    var mtmp0 Tracker = self__0
    var x1 string = mtmp0.label
    var x2 *ref_int_x = mtmp0.count
    var x3 *ref_bool_x = mtmp0.toggled
    var toggled__3 *ref_bool_x = x3
    var count__2 *ref_int_x = x2
    var label__1 string = x1
    var current__4 int = ref_get__Ref_int(count__2)
    var flag__5 bool = ref_get__Ref_bool(toggled__3)
    var with_label__6 string = string_add("Tracker(", label__1)
    var with_count_label__7 string = string_add(with_label__6, ", count: ")
    var t38 string = int_to_string(current__4)
    var with_count__8 string = string_add(with_count_label__7, t38)
    var with_flag_label__9 string = string_add(with_count__8, ", toggled: ")
    var t40 string = bool_to_string(flag__5)
    var t39 string = string_add(t40, ")")
    ret64 = string_add(with_flag_label__9, t39)
    return ret64
}

func impl_Describe_Record_int_describe(self__10 Record__int) string {
    var ret65 string
    switch self__10 := self__10.(type) {
    case Record__int_Value:
        var x4 int = self__10._0
        var value__11 int = x4
        var t42 string = int_to_string(value__11)
        var t41 string = string_add(t42, ")")
        ret65 = string_add("Value(", t41)
    case Record__int_Pair:
        var x5 int = self__10._0
        var x6 int = self__10._1
        var after__13 int = x6
        var before__12 int = x5
        var t43 string = int_to_string(before__12)
        var prefix__14 string = string_add("Pair(", t43)
        var t46 string = int_to_string(after__13)
        var t45 string = string_add(t46, ")")
        var t44 string = string_add(", ", t45)
        ret65 = string_add(prefix__14, t44)
    case Record__int_Empty:
        ret65 = "Empty"
    }
    return ret65
}

func impl_Describe_Record_string_describe(self__15 Record__string) string {
    var ret66 string
    switch self__15 := self__15.(type) {
    case Record__string_Value:
        var x7 string = self__15._0
        var text__16 string = x7
        var t47 string = string_add(text__16, ")")
        ret66 = string_add("Value(", t47)
    case Record__string_Pair:
        var x8 string = self__15._0
        var x9 string = self__15._1
        var after__18 string = x9
        var before__17 string = x8
        var prefix__19 string = string_add("Pair(", before__17)
        var t49 string = string_add(after__18, ")")
        var t48 string = string_add(", ", t49)
        ret66 = string_add(prefix__19, t48)
    case Record__string_Empty:
        ret66 = "Empty"
    }
    return ret66
}

func format_total(total__26 int) string {
    var ret67 string
    var t50 string = int_to_string(total__26)
    ret67 = string_add("total: ", t50)
    return ret67
}

func increment(value__27 int) int {
    var ret68 int
    ret68 = int_add(value__27, 1)
    return ret68
}

func triple(value__28 int) int {
    var ret69 int
    ret69 = int_mul(value__28, 3)
    return ret69
}

func pair_join(parts__29 Tuple2_string_string) string {
    var ret70 string
    var mtmp11 Tuple2_string_string = parts__29
    var x12 string = mtmp11._0
    var x13 string = mtmp11._1
    var right__31 string = x13
    var left__30 string = x12
    var t51 string = string_add(" -> ", right__31)
    ret70 = string_add(left__30, t51)
    return ret70
}

func run_transforms(value__32 int, transforms__33 [2]func(int) int) [2]int {
    var ret71 [2]int
    var first__34 func(int) int = array_get__Array_2_Fn_int_to_int(transforms__33, 0)
    var second__35 func(int) int = array_get__Array_2_Fn_int_to_int(transforms__33, 1)
    var first_result__36 int = first__34(value__32)
    var second_result__37 int = second__35(first_result__36)
    var t52 [2]int = [2]int{first_result__36, value__32}
    ret71 = array_set__Array_2_int(t52, 1, second_result__37)
    return ret71
}

func gather(record__38 Record__int) Maybe__int {
    var ret72 Maybe__int
    switch record__38 := record__38.(type) {
    case Record__int_Value:
        var x14 int = record__38._0
        var value__39 int = x14
        ret72 = Maybe__int_Some{
            _0: value__39,
        }
    case Record__int_Pair:
        var x16 int = record__38._1
        var after__40 int = x16
        ret72 = Maybe__int_Some{
            _0: after__40,
        }
    case Record__int_Empty:
        ret72 = Maybe__int_None{}
    }
    return ret72
}

func build_counter(label__41 string, start__42 int) Tuple4_Tracker_closure_env_snapshot_0_closure_env_bump_1_closure_env_flip_2 {
    var ret73 Tuple4_Tracker_closure_env_snapshot_0_closure_env_bump_1_closure_env_flip_2
    var count__43 *ref_int_x = ref__Ref_int(start__42)
    var toggled__44 *ref_bool_x = ref__Ref_bool(false)
    var tracker__45 Tracker = Tracker{
        label: label__41,
        count: count__43,
        toggled: toggled__44,
    }
    var snapshot__46 closure_env_snapshot_0 = closure_env_snapshot_0{
        count_0: count__43,
    }
    var bump__49 closure_env_bump_1 = closure_env_bump_1{
        count_0: count__43,
    }
    var flip__52 closure_env_flip_2 = closure_env_flip_2{
        toggled_0: toggled__44,
    }
    ret73 = Tuple4_Tracker_closure_env_snapshot_0_closure_env_bump_1_closure_env_flip_2{
        _0: tracker__45,
        _1: snapshot__46,
        _2: bump__49,
        _3: flip__52,
    }
    return ret73
}

func main0() struct{} {
    var ret74 struct{}
    var mtmp19 Tuple4_Tracker_closure_env_snapshot_0_closure_env_bump_1_closure_env_flip_2 = build_counter("goml", 2)
    var x20 Tracker = mtmp19._0
    var x21 closure_env_snapshot_0 = mtmp19._1
    var x22 closure_env_bump_1 = mtmp19._2
    var x23 closure_env_flip_2 = mtmp19._3
    var flip__56 closure_env_flip_2 = x23
    var bump__55 closure_env_bump_1 = x22
    var snapshot__54 closure_env_snapshot_0 = x21
    var tracker__53 Tracker = x20
    var tracker_info__57 string = impl_Describe_Tracker_describe(tracker__53)
    var first_record__58 Record__int = __closure_apply32(snapshot__54)
    var bumped_record__59 Record__int = __closure_apply34(bump__55, 5)
    var flipped_record__60 Record__string = __closure_apply36(flip__56)
    var maybe_first__61 Maybe__int = gather(first_record__58)
    var maybe_second__62 Maybe__int = gather(bumped_record__59)
    var chosen__63 Maybe__int = choose__T_Maybe_int(true, maybe_second__62, maybe_first__61)
    var stringified__64 Maybe__string = map_maybe__T_int__U_string(chosen__63, format_total)
    var transforms__65 [2]func(int) int = [2]func(int) int{increment, triple}
    var results__66 [2]int = run_transforms(4, transforms__65)
    var first_result__67 int = array_get__Array_2_int(results__66, 0)
    var second_result__68 int = array_get__Array_2_int(results__66, 1)
    var t53 bool = int_less(first_result__67, second_result__68)
    var order_check__69 bool = bool_and(t53, true)
    var first_text__70 string = impl_Describe_Record_int_describe(first_record__58)
    var bumped_text__71 string = impl_Describe_Record_int_describe(bumped_record__59)
    var flipped_text__72 string = impl_Describe_Record_string_describe(flipped_record__60)
    var summary__74 string
    switch stringified__64 := stringified__64.(type) {
    case Maybe__string_Some:
        var x24 string = stringified__64._0
        var text__73 string = x24
        summary__74 = string_add("Snapshot: ", text__73)
    case Maybe__string_None:
        summary__74 = "Snapshot: none"
    }
    var t55 string = int_to_string(first_result__67)
    var t56 string = int_to_string(second_result__68)
    var t54 Tuple2_string_string = Tuple2_string_string{
        _0: t55,
        _1: t56,
    }
    var pair_text__75 string = pair_join(t54)
    var bool_text__76 string = bool_to_string(order_check__69)
    string_println(tracker_info__57)
    string_println(first_text__70)
    string_println(bumped_text__71)
    string_println(flipped_text__72)
    string_println(summary__74)
    string_println(pair_text__75)
    string_println(bool_text__76)
    ret74 = struct{}{}
    return ret74
}

func __closure_apply32(env33 closure_env_snapshot_0) Record__int {
    var ret75 Record__int
    var count__43 *ref_int_x = env33.count_0
    var t57 int = ref_get__Ref_int(count__43)
    ret75 = Record__int_Value{
        _0: t57,
    }
    return ret75
}

func __closure_apply34(env35 closure_env_bump_1, delta__47 int) Record__int {
    var ret76 Record__int
    var count__43 *ref_int_x = env35.count_0
    var before__48 int = ref_get__Ref_int(count__43)
    var t58 int = int_add(before__48, delta__47)
    ref_set__Ref_int(count__43, t58)
    var t59 int = ref_get__Ref_int(count__43)
    ret76 = Record__int_Pair{
        _0: before__48,
        _1: t59,
    }
    return ret76
}

func __closure_apply36(env37 closure_env_flip_2) Record__string {
    var ret77 Record__string
    var toggled__44 *ref_bool_x = env37.toggled_0
    var before__50 bool = ref_get__Ref_bool(toggled__44)
    var t60 bool = bool_not(before__50)
    ref_set__Ref_bool(toggled__44, t60)
    var after__51 bool = ref_get__Ref_bool(toggled__44)
    var t61 string = bool_to_string(before__50)
    var t62 string = bool_to_string(after__51)
    ret77 = Record__string_Pair{
        _0: t61,
        _1: t62,
    }
    return ret77
}

func choose__T_Maybe_int(flag__20 bool, when_true__21 Maybe__int, when_false__22 Maybe__int) Maybe__int {
    var ret78 Maybe__int
    if flag__20 {
        ret78 = when_true__21
    } else {
        ret78 = when_false__22
    }
    return ret78
}

func map_maybe__T_int__U_string(value__23 Maybe__int, f__24 func(int) string) Maybe__string {
    var ret79 Maybe__string
    switch value__23 := value__23.(type) {
    case Maybe__int_Some:
        var x10 int = value__23._0
        var inner__25 int = x10
        var t63 string = f__24(inner__25)
        ret79 = Maybe__string_Some{
            _0: t63,
        }
    case Maybe__int_None:
        ret79 = Maybe__string_None{}
    }
    return ret79
}

func main() {
    main0()
}
