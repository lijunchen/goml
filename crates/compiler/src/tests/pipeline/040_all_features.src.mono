fn impl_Describe_Tracker_describe(self/0: Tracker) -> string {
  let mtmp0 = self/0 in
  let x1 = Tracker.label(mtmp0) in
  let x2 = Tracker.count(mtmp0) in
  let x3 = Tracker.toggled(mtmp0) in
  let toggled/3 = x3 in
  let count/2 = x2 in
  let label/1 = x1 in
  let current/4 = ref_get(count/2) in
  let flag/5 = ref_get(toggled/3) in
  let with_label/6 = string_add("Tracker(", label/1) in
  let with_count_label/7 = string_add(with_label/6, ", count: ") in
  let with_count/8 = string_add(with_count_label/7, int_to_string(current/4)) in
  let with_flag_label/9 = string_add(with_count/8, ", toggled: ") in
  string_add(with_flag_label/9, string_add(bool_to_string(flag/5), ")"))
}

fn impl_Describe_Record_int_describe(self/10: Record__int) -> string {
  match self/10 {
    Record__int::Value(x4) => {
      let x4 = Record__int::Value._0(self/10) in
      let value/11 = x4 in
      string_add("Value(", string_add(int_to_string(value/11), ")"))
    },
    Record__int::Pair(x5, x6) => {
      let x5 = Record__int::Pair._0(self/10) in
      let x6 = Record__int::Pair._1(self/10) in
      let after/13 = x6 in
      let before/12 = x5 in
      let prefix/14 = string_add("Pair(", int_to_string(before/12)) in
      string_add(prefix/14, string_add(", ", string_add(int_to_string(after/13), ")")))
    },
    Record__int::Empty => {
      "Empty"
    },
  }
}

fn impl_Describe_Record_string_describe(self/15: Record__string) -> string {
  match self/15 {
    Record__string::Value(x7) => {
      let x7 = Record__string::Value._0(self/15) in
      let text/16 = x7 in
      string_add("Value(", string_add(text/16, ")"))
    },
    Record__string::Pair(x8, x9) => {
      let x8 = Record__string::Pair._0(self/15) in
      let x9 = Record__string::Pair._1(self/15) in
      let after/18 = x9 in
      let before/17 = x8 in
      let prefix/19 = string_add("Pair(", before/17) in
      string_add(prefix/19, string_add(", ", string_add(after/18, ")")))
    },
    Record__string::Empty => {
      "Empty"
    },
  }
}

fn format_total(total/26: int) -> string {
  string_add("total: ", int_to_string(total/26))
}

fn increment(value/27: int) -> int {
  int_add(value/27, 1)
}

fn triple(value/28: int) -> int {
  int_mul(value/28, 3)
}

fn pair_join(parts/29: (string, string)) -> string {
  let mtmp11 = parts/29 in
  let x12 = mtmp11.0 in
  let x13 = mtmp11.1 in
  let right/31 = x13 in
  let left/30 = x12 in
  string_add(left/30, string_add(" -> ", right/31))
}

fn run_transforms(value/32: int, transforms/33: [(int) -> int; 2]) -> [int; 2] {
  let first/34 = array_get(transforms/33, 0) in
  let second/35 = array_get(transforms/33, 1) in
  let first_result/36 = first/34(value/32) in
  let second_result/37 = second/35(first_result/36) in
  array_set([first_result/36, value/32], 1, second_result/37)
}

fn gather(record/38: Record__int) -> Maybe__int {
  match record/38 {
    Record__int::Value(x14) => {
      let x14 = Record__int::Value._0(record/38) in
      let value/39 = x14 in
      Maybe__int::Some(value/39)
    },
    Record__int::Pair(x15, x16) => {
      let x15 = Record__int::Pair._0(record/38) in
      let x16 = Record__int::Pair._1(record/38) in
      let after/40 = x16 in
      Maybe__int::Some(after/40)
    },
    Record__int::Empty => {
      Maybe__int::None
    },
  }
}

fn build_counter(label/41: string, start/42: int) -> (Tracker, closure_env_snapshot_0, closure_env_bump_1, closure_env_flip_2) {
  let count/43 = ref(start/42) in
  let toggled/44 = ref(false) in
  let tracker/45 = Tracker { label: label/41, count: count/43, toggled: toggled/44 } in
  let snapshot/46 = closure_env_snapshot_0 { count_0: count/43 } in
  let bump/49 = closure_env_bump_1 { count_0: count/43 } in
  let flip/52 = closure_env_flip_2 { toggled_0: toggled/44 } in
  (tracker/45, snapshot/46, bump/49, flip/52)
}

fn main() -> unit {
  let mtmp19 = build_counter("goml", 2) in
  let x20 = mtmp19.0 in
  let x21 = mtmp19.1 in
  let x22 = mtmp19.2 in
  let x23 = mtmp19.3 in
  let flip/56 = x23 in
  let bump/55 = x22 in
  let snapshot/54 = x21 in
  let tracker/53 = x20 in
  let tracker_info/57 = impl_Describe_Tracker_describe(tracker/53) in
  let first_record/58 = __closure_apply32(snapshot/54) in
  let bumped_record/59 = __closure_apply34(bump/55, 5) in
  let flipped_record/60 = __closure_apply36(flip/56) in
  let maybe_first/61 = gather(first_record/58) in
  let maybe_second/62 = gather(bumped_record/59) in
  let chosen/63 = choose__T_Maybe_int(true, maybe_second/62, maybe_first/61) in
  let stringified/64 = map_maybe__T_int__U_string(chosen/63, format_total) in
  let transforms/65 = [increment, triple] in
  let results/66 = run_transforms(4, transforms/65) in
  let first_result/67 = array_get(results/66, 0) in
  let second_result/68 = array_get(results/66, 1) in
  let order_check/69 = bool_and(int_less(first_result/67, second_result/68), true) in
  let first_text/70 = impl_Describe_Record_int_describe(first_record/58) in
  let bumped_text/71 = impl_Describe_Record_int_describe(bumped_record/59) in
  let flipped_text/72 = impl_Describe_Record_string_describe(flipped_record/60) in
  let summary/74 = match stringified/64 {
    Maybe__string::Some(x24) => {
      let x24 = Maybe__string::Some._0(stringified/64) in
      let text/73 = x24 in
      string_add("Snapshot: ", text/73)
    },
    Maybe__string::None => {
      "Snapshot: none"
    },
  } in
  let pair_text/75 = pair_join((int_to_string(first_result/67), int_to_string(second_result/68))) in
  let bool_text/76 = bool_to_string(order_check/69) in
  let mtmp25 = string_println(tracker_info/57) in
  let mtmp26 = string_println(first_text/70) in
  let mtmp27 = string_println(bumped_text/71) in
  let mtmp28 = string_println(flipped_text/72) in
  let mtmp29 = string_println(summary/74) in
  let mtmp30 = string_println(pair_text/75) in
  let mtmp31 = string_println(bool_text/76) in
  ()
}

fn __closure_apply32(env33: closure_env_snapshot_0) -> Record__int {
  let count/43 = closure_env_snapshot_0.count_0(env33) in
  Record__int::Value(ref_get(count/43))
}

fn __closure_apply34(env35: closure_env_bump_1, delta/47: int) -> Record__int {
  let count/43 = closure_env_bump_1.count_0(env35) in
  let before/48 = ref_get(count/43) in
  let mtmp17 = ref_set(count/43, int_add(before/48, delta/47)) in
  Record__int::Pair(before/48, ref_get(count/43))
}

fn __closure_apply36(env37: closure_env_flip_2) -> Record__string {
  let toggled/44 = closure_env_flip_2.toggled_0(env37) in
  let before/50 = ref_get(toggled/44) in
  let mtmp18 = ref_set(toggled/44, bool_not(before/50)) in
  let after/51 = ref_get(toggled/44) in
  Record__string::Pair(bool_to_string(before/50), bool_to_string(after/51))
}

fn choose__T_Maybe_int(flag/20: bool, when_true/21: Maybe__int, when_false/22: Maybe__int) -> Maybe__int {
  if flag/20 {
      when_true/21
  } else {
      when_false/22
  }
}

fn map_maybe__T_int__U_string(value/23: Maybe__int, f/24: (int) -> string) -> Maybe__string {
  match value/23 {
    Maybe__int::Some(x10) => {
      let x10 = Maybe__int::Some._0(value/23) in
      let inner/25 = x10 in
      Maybe__string::Some(f/24(inner/25))
    },
    Maybe__int::None => {
      Maybe__string::None
    },
  }
}