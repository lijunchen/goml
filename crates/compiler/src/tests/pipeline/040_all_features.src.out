Tracker(goml, count: 2, toggled: false)
Value(2)
Pair(2, 7)
Pair(false, true)
Snapshot: total: 7
5 -> 15
true
